### PR TITLE
Create -Xtune:throughput option and set options that favor throughput plus other cleanup

### PR DESCRIPTION
**Pull request by mpirvu**
This PR creates a new option, called `-Xtune:throughput`, under which the JVM will favor throughput over other performance metrics. In particular, if not explicitly provided, the initial heap size (`-Xms`) will be set to the maximum heap size value (`-Xmx`). Moreover, the JIT will internally set some options to optimize more aggressively the code.

Another new option is `-Xtune:quickstart` which is set to be an alias of `-Xquickstart`.
The 3 options `-Xtune:virtualized`, `-Xtune:throughput` and `-Xtune:quickstart` are mutually exclusive (since they pull performance metrics in different directions) and the last occurrence on the command line takes effect (this is different than the old implementation where quickstart mode had precedence over -Xtune:virtualized).

The JIT code that was parsing `-server` and `-client` options has been removed since it wasn't working anyway. 

Depends on https://github.com/eclipse/omr/pull/6294

Issue: #14246

### CODE REVIEW

**Comment by mpirvu** [comment]
@amicic or @dmitripivkine Could you please review the changes relevant to GC? Thanks

**Comment by mpirvu** [comment]
FYI: @vijaysun-omr

File: runtime/gc_base/GCExtensions.cpp [hunk 1]
```diff
@@ -235,7 +235,7 @@ MM_GCExtensions::updateIdentityHashDataForSaltIndex(UDATA index)
 }
 
 /*
- * computeDefaultMaxHeapForJava is for Java only, it will be called during gcParseCommandLineAndInitializeWithValues(),
+ * computeDefaultMaxHeapForJava is for Java8 only, it will be called during gcParseCommandLineAndInitializeWithValues(),
```
Comment (dmitripivkine) [inline runtime/gc_base/GCExtensions.cpp:238]: - I don't understand reason for this change, it is not accurate to say this function set values for Java 8 only.
- As fas as reminding changes in this file are white spaces only this file should be excluded from this change set and not touched at all

Comment (mpirvu) [reply runtime/gc_base/GCExtensions.cpp]: I have excluded the file from the commit.

File: runtime/gc_modron_startup/mminit.cpp [hunk 2]
```diff
@@ -795,11 +795,17 @@ gcInitializeCalculatedValues(J9JavaVM *javaVM, IDATA* memoryParameters)
...
+	UDATA initialXmsValueMax = (8 * 1024 * 1024);
+	UDATA initialXmsValueMin = (8 * 1024 * 1024);
+	if (extensions->isSegregatedHeap() || extensions->isMetronomeGC()) {
+		/* TODO aryoung: eventually segregated heaps will allow heap expansion, although metronome
+		 * itself will still require a fully expanded heap on startup */
+		initialXmsValueMax = J9_MEMORY_MAX;
+		initialXmsValueMin = UDATA_MAX;
+	} else if (J9_ARE_ANY_BITS_SET(javaVM->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_TUNE_THROUGHPUT)) {
+		/* For -Xtune:throughput we want to set Xms=Xmx */
+		initialXmsValueMax = initialXmsValueMin = extensions->memoryMax;
```
Comment (dmitripivkine) [inline runtime/gc_modron_startup/mminit.cpp:807]: Multiple assignments in the same operator are not allowed by Coding Standards. Please split this to two operators.

Comment (dmitripivkine) [reply runtime/gc_modron_startup/mminit.cpp:807]: Also would you please double check this code works if `-Xmx` value is provided explicitly (both cases larger and smaller than JVM default - 25% of RAM for Java 11 and up or 512m for Java 8)?

**Review by dmitripivkine — CHANGES_REQUESTED** [review]

**Comment by mpirvu** [comment]
FYI @keithc-ca for changes done  to jvminit.h/c

**Review by dmitripivkine — APPROVED** [review]
GC change looks good for me

File: runtime/oti/jvminit.h [hunk 3]
```diff
@@ -426,6 +426,8 @@ enum INIT_STAGE {
 #define VMOPT_XXNODISCLAIMJITSCRATCH	"-XX:-DisclaimJitScratch"
 
 #define VMOPT_TUNE_VIRTUALIZED "-Xtune:virtualized"
+#define VMOPT_TUNE_THROUGHPUT "-Xtune:throughput"
+#define VMOPT_TUNE_QUICKSTART "-Xtune:quickstart"
```
Comment (keithc-ca) [inline runtime/oti/jvminit.h:430]: Suggest sorting these by name.

File: runtime/vm/jvminit.c [hunk 4]
```diff
@@ -1994,9 +1991,34 @@ VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved)
 			}
 #endif
 
+			IDATA xtuneQuickstartIndex =  FIND_AND_CONSUME_ARG(EXACT_MATCH, VMOPT_TUNE_QUICKSTART, NULL);
+			IDATA xquickstartIndex =  FIND_AND_CONSUME_ARG(EXACT_MATCH, VMOPT_XQUICKSTART, NULL);
+			IDATA clientIndex =  FIND_AND_CONSUME_ARG(EXACT_MATCH, VMOPT_CLIENT, NULL);
```
Comment (keithc-ca) [inline runtime/vm/jvminit.c:1996]: nit: please use a single space after each `=`

File: runtime/vm/jvminit.c [hunk 5]
```diff
@@ -1994,9 +1991,34 @@ VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved)
...
+			/* VMOPT_TUNE_QUICKSTART, VMOPT_XQUICKSTART and VMOPT_CLIENT are
+			 * aliases, so take the largest index of these 3 */
+			if (xquickstartIndex > xtuneQuickstartIndex) {
+				xtuneQuickstartIndex = xquickstartIndex;
+			}
+			if (clientIndex > xtuneQuickstartIndex) {
+				xtuneQuickstartIndex = clientIndex;
+			}
```
Comment (keithc-ca) [inline runtime/vm/jvminit.c:2004]: It's easier to see that the names being compared are the same used in the assignment if they appear in the same order:
```
	if (xtuneQuickstartIndex < xquickstartIndex) {
		xtuneQuickstartIndex = xquickstartIndex;
	}
	if (xtuneQuickstartIndex < clientIndex) {
		xtuneQuickstartIndex = clientIndex;
	}
```

File: runtime/oti/jvminit.h (hunk 3 above)
Comment (mpirvu) [reply runtime/oti/jvminit.h]: Done.

File: runtime/vm/jvminit.c (hunk 4 above)
Comment (mpirvu) [reply runtime/vm/jvminit.c]: Done.

File: runtime/vm/jvminit.c (hunk 5 above)
Comment (mpirvu) [reply runtime/vm/jvminit.c]: Done.

File: runtime/gc_modron_startup/mminit.cpp [hunk 6]
```diff
@@ -795,11 +795,18 @@ gcInitializeCalculatedValues(J9JavaVM *javaVM, IDATA* memoryParameters)
 	 * This used to be for free, calculations based on Xmx, now it is a manual check
 	 * in setConfigurationSpecificMemoryParameters
 	 */
-	/* TODO aryoung: eventually segregated heaps will allow heap expansion, although metronome
-	 * itself will still require a fully expanded heap on startup */
-	const bool shouldMaximizeInitialHeap = (extensions->isSegregatedHeap() || extensions->isMetronomeGC());
-	const UDATA initialXmsValueMax = shouldMaximizeInitialHeap ? J9_MEMORY_MAX : (8 * 1024 * 1024);
-	const UDATA initialXmsValueMin = shouldMaximizeInitialHeap ? UDATA_MAX     : (8 * 1024 * 1024);
+	UDATA initialXmsValueMax = (8 * 1024 * 1024);
+	UDATA initialXmsValueMin = (8 * 1024 * 1024);
```
Comment (keithc-ca) [inline runtime/gc_modron_startup/mminit.cpp:799]: The parentheses here seem redundant.

File: runtime/gc_modron_startup/mminit.cpp [hunk 7]
```diff
@@ -795,11 +795,18 @@ gcInitializeCalculatedValues(J9JavaVM *javaVM, IDATA* memoryParameters)
...
+	if (extensions->isSegregatedHeap() || extensions->isMetronomeGC()) {
+		/* TODO aryoung: eventually segregated heaps will allow heap expansion, although metronome
+		 * itself will still require a fully expanded heap on startup */
```
Comment (keithc-ca) [inline runtime/gc_modron_startup/mminit.cpp:802]: Multi-line comments should have closing `*/` on a separate line.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: runtime/compiler/control/DLLMain.cpp [hunk 8]
```diff
@@ -138,16 +136,7 @@ IDATA J9VMDllMain(J9JavaVM* vm, IDATA stage, void * reserved)
...
-         // on the command line. The last appearance wins
-         if (argIndexServer < argIndexClient || argIndexServer < argIndexQuickstart)
-            {
-            isQuickstart = true;
-            }
-         else if (argIndexServer > 0)
-            {
-            TR::Options::_bigAppThreshold = 1;
-            }
+         isQuickstart = J9_ARE_ALL_BITS_SET(vm->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_TUNE_QUICKSTART);
```
[the remark this answers is not in the document]
Comment (mpirvu) [reply runtime/compiler/control/DLLMain.cpp]: Done

File: runtime/compiler/control/J9Options.cpp [hunk 9]
```diff
@@ -1305,40 +1305,37 @@ void J9::Options::preProcessMmf(J9JavaVM *vm, J9JITConfig *jitConfig)
...
       {
-      // Xtune:virtualized will put us in aggressivenessLevel3, but only if other options
-      // like Xquickstart, -client, -server, -Xaggressive are not specified
-      if (vm->runtimeFlags & J9_RUNTIME_TUNE_VIRTUALIZED)
+      // Start with a default level and override as needed
+      _aggressivenessLevel = TR::Options::TR_AggresivenessLevel::DEFAULT;
+
+      // -Xquickstart/-Xtune:quickstart, -Xtune:virtualized and -Xtune:throughput are mutually exclusive
+      // This is ensured by VM option processing
+      if (jitConfig->runtimeFlags & J9JIT_QUICKSTART)
```
[the remark this answers is not in the document]
Comment (mpirvu) [reply runtime/compiler/control/J9Options.cpp]: Done

File: runtime/gc_modron_startup/mminit.cpp (hunk 6 above)
Comment (mpirvu) [reply runtime/gc_modron_startup/mminit.cpp]: Done

File: runtime/gc_modron_startup/mminit.cpp (hunk 7 above)
Comment (mpirvu) [reply runtime/gc_modron_startup/mminit.cpp]: Done

**Comment by mpirvu** [comment]
OMR dependency has been promoted, so this PR can be delivered.

**Comment by mpirvu** [comment]
jenkins compile all jdk8

**Comment by mpirvu** [comment]
@keithc-ca if there are no more suggestions to be implemented, could you please approve the PR? Thanks!

**Review by keithc-ca — APPROVED** [review]

**Comment by keithc-ca** [comment]
jenkins test sanity alinux64 jdk17

**Comment by vijaysun-omr** [comment]
Is it possible to update the PR title, abstract and the different commit messages to capture all of the changes being made ? I looked at the JIT commit and it was doing 3-4 things not all of which seem to be captured in all those relevant places.

**Comment by keithc-ca** [comment]
For the record, the testing passed:
https://openj9-jenkins.osuosl.org/view/Pull%20Requests/job/PullRequest-OpenJ9/1417/

**Comment by mpirvu** [comment]
I have updated the title and commit message for the JIT commit (https://github.com/eclipse-openj9/openj9/pull/14277/commits/d0c8af76164af658641f95d9108fb3d2a6ad522b) to be more descriptive.
I have also slightly changed the title and description of this PR.

**Comment by vijaysun-omr** [comment]
Thanks, `precendence` is a typo in the new commit message.

**Comment by vijaysun-omr** [comment]
`apperance` is another typo

**Comment by vijaysun-omr** [comment]
Since the prior testing did pass, and I have no issues with the code changes, and we are only updating commit/PR messages here, I am thinking of merging it after the updates to messages are done.

**Comment by mpirvu** [comment]
I have corrected the typos.

> Just wanted to check if the behavior of the aliases -Xtune:quickstart and -Xquickstart is consistent, e.g. when compared to the precedence vs -Xtune:virtualized.

The current code does not use any "precedence" rules. The last option wins.  -Xtune:quickstart and -Xquickstart are considered aliases and you can use them interchangeably regardless of the position on the command line.
In a future PR I plan to add information in the verbose logs about the mode the JVM is in (quickstart, throughput or virtualized)

**Review by vijaysun-omr — APPROVED** [review]

**Pull request merged**